src/builtins.c

```diff
@@ -1075,6 +1079,49 @@ JL_CALLABLE(jl_f_arrayset)
     return args[1];
 }
 
+JL_CALLABLE(jl_f_arrayfreeze)
+{
+    JL_NARGSV(arrayfreeze, 1);
+    JL_TYPECHK(arrayfreeze, array, args[0]);
+    jl_array_t *a = (jl_array_t*)args[0];
+    jl_datatype_t *it = (jl_datatype_t *)jl_apply_type2((jl_value_t*)jl_immutable_array_type,
+        jl_tparam0(jl_typeof(a)), jl_tparam1(jl_typeof(a)));
+    // The idea is to elide this copy if the compiler or runtime can prove that
+    // doing so is safe to do.
+    jl_array_t *na = jl_array_copy(a);
+    jl_set_typeof(na, it);
+    return (jl_value_t*)na;
+}
+
+JL_CALLABLE(jl_f_mutating_arrayfreeze)
```

Comment:
At first, I was thrown by what this name meant. It makes sense in the end; in this case you are sort-of mutating the type of `a` but not the value/data... though I did wonder if it would be better language to describe this as "taking" or "stealing" `a` or something like that.